google-cloud-bigtable/lib/google/cloud/bigtable/table_data_operations.rb

```diff
@@ -0,0 +1,483 @@
+          @app_profile_id = app_profile_id
+        end
+
+        # Read rows
+        #
+        # @param rows [Google::Bigtable::V2::RowSet | Hash]
+        #   The row keys and/or ranges to read.
+        #   If not specified, reads from all rows.
+        #   A hash of the same form as `Google::Bigtable::V2::RowSet`
+        #   can also be provided.
+        # @param filter [Google::Bigtable::V2::RowFilter | Hash]
+        #   The filter to apply to the contents of the specified row(s). If unset,
+        #   reads the entirety of each row.
+        #   A hash of the same form as `Google::Bigtable::V2::RowFilter`
+        #   can also be provided.
+        # @param rows_limit [Integer]
+        #   The read will terminate after committing to N rows' worth of results.
+        #   The default (zero) is to return all results.
+        # @param options [Google::Gax::CallOptions]
+        #   Overrides the default settings for this call, e.g, timeout,
+        #   retries, etc.
+        # @return [Array<Google::Cloud::Bigtable::FlatRow> | :yields: row]
+        #   Array of row or yield block for each processed row.
+        # @raise [Google::Gax::GaxError] if the RPC is aborted.
+        # @example
+        #   require "google/cloud/bigtable"
+        #
+        #   client = Google::Cloud::Bigtable.new(instance_id: "instance-id")
+        #
+        #   table = client.table("table-name")
+        #
+        #   table.read_rows(rows_limit: 100) do |row|
+        #     p row
+        #   end
+        #
+        #   # OR - without block
+        #   rows = table.read_rows(rows_limit: 100)
+        #
+        #   # With row keys
+        #   rows_set = Google::Bigtable::V2::RowSet.new
+        #   row_set.row_keys << "user01"
+        #   row_set.row_keys << "user02"
+        #
+        #   row_range = Google::Bigtable::V2::RowRange.new({
+        #     start_key_closed: "user00",
+        #     end_key_closed: "user10"
+        #   )
+
+        #   row_set.row_ranges << row_range
+        #
+        #   table.read_rows(rows: rows_set) do |row|
+        #     p row
+        #   end
+
+        def read_rows \
+            rows: nil,
+            filter: nil,
+            rows_limit: nil,
+            options: nil,
+            &block
+          max_retries = DEFAULT_READ_RETRY_COUNT
+          if options && options[:max_retries].to_i > 0
+            max_retries = options[:max_retries]
+          end
+
+          retry_count = 0
+          req_row_set = build_row_set(rows)
+          req_rows_limit = rows_limit
+          rows_reader = RowsReader.new(
+            client,
+            table_path,
+            app_profile_id,
+            options
+          )
+
+          begin
+            retry_count = 0
+            rows_reader.read(
+              rows: req_row_set,
+              filter: filter,
+              rows_limit: req_rows_limit,
+              &block
+            )
+          rescue *GRPC_RETRYABLE_ERRORS => e
```

Comment:
Why are retries being handled here? Shouldn't this already be handled by the GAPIC code and GAX?

Reply:
Yes, GAX is able to do retry but in a read rows case needed special kind of retry mechanism.

On `GRPC_RETRYABLE_ERRORS ` errors starts reading from last read keys or by updating read options.

i.e in `rows_limit`, `row key ranges`, an array of `row keys` options provided then retry read on
 error by changing provided options(update rows limit, row range, removed already read row keys from row keys)

Reply:
Makes sense. Thanks. We do something similar in Spanner and Firestore.